### PR TITLE
Fix broken intra-doc links + add CI stage to test example docs

### PR DESCRIPTION
**Pull request by cmichi**
You can verify for yourself by executing

* `cargo doc --workspace --document-private-items --verbose --no-deps` in `ink/` or

* `cargo doc --manifest-path=Cargo.toml --document-private-items --verbose --no-deps` in an example folder.

During the issue I discovered [a bug in rustdoc](https://github.com/rust-lang/rust/issues/79792).

### CODE REVIEW

File: crates/lang/codegen/src/generator/cross_calling.rs [hunk 1]
```diff
@@ -140,7 +140,7 @@ impl CrossCalling<'_> {
         }
     }
 
-    /// Builds up the [`ink_env::call::ArgumentList`] type structure for the given types.
+    /// Builds up the `ink_env::call::utils::ArgumentList` type structure for the given types.
```
Comment (cmichi) [inline crates/lang/codegen/src/generator/cross_calling.rs:143]: `ink_env` is no dependency of this crate, hence intra-crate links to `ink_env` do not work.

Making it a dependency for this purpose seems kinda heavy and would also not work (I tried it in [this CI run](https://gitlab.parity.io/parity/ink/-/pipelines/116245)).

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -264,6 +264,20 @@ examples-contract-build:
...
+    <<:                              *docker-env
+    variables:
+        RUSTDOCFLAGS: -Dwarnings
+    script:
+        # `--document-private-items` needs to be in here because currently our contract macro
+        # puts the contract functions in a private module.
+        # Once https://github.com/paritytech/ink/issues/336 has been implemented we can get rid
+        # of this flag.
+        - for example in examples/*/; do
+            cargo doc --manifest-path ${example}/Cargo.toml --document-private-items --verbose --no-deps;
```
Comment (Robbepop) [inline .gitlab-ci.yml:278]: last time I tried to use `--no-deps` it actually compiled docs for dependencies and even reported errors and warnings for them. did you check that this works?

File: crates/storage/src/collections/stash/mod.rs [hunk 3]
```diff
@@ -33,6 +33,9 @@ use crate::{
 };
 use ink_primitives::Key;
 
+#[cfg(doc)]
+use crate::traits::SpreadLayout;
+
```
Comment (Robbepop) [inline crates/storage/src/collections/stash/mod.rs:38]: Why is this needed in `cfg(doc)`?

File: crates/storage/src/lazy/lazy_hmap.rs [hunk 4]
```diff
@@ -54,6 +54,9 @@ use ink_prelude::{
 };
 use ink_primitives::Key;
 
+#[cfg(doc)]
+use crate::lazy::LazyIndexMap;
+
```
Comment (Robbepop) [inline crates/storage/src/lazy/lazy_hmap.rs:59]: Why is this needed?

File: .gitlab-ci.yml (hunk 2 above)
Comment (cmichi) [reply .gitlab-ci.yml:278]: Yes I tried and it works now.

File: crates/storage/src/collections/stash/mod.rs (hunk 3 above)
Comment (cmichi) [reply crates/storage/src/collections/stash/mod.rs:38]: Needed for the crate documentation of `storage`:
```
ink/crates/storage
cargo doc --document-private-items --no-deps
```

Thing is that it's only needed for `--document-private-items`, so not sure which policy is best suited for that. IMHO we should ensure that all doc links are working, no matter if private or public. On the other hand we only test the public interface for the crates in CI currently. Wdyt?

File: crates/storage/src/lazy/lazy_hmap.rs (hunk 4 above)
Comment (cmichi) [reply crates/storage/src/lazy/lazy_hmap.rs:59]: Same as above.

Comment (Robbepop) [reply crates/storage/src/lazy/lazy_hmap.rs:59]: I think I'd rather want to write-out the path entirely for the private docs instead of having another import statement that is potentially confusing other people just as me.

**Comment by cmichi** [comment]
On my machine the latest `rustc` nightly (from yesterday) results in this ICE when compiling: https://github.com/rust-lang/rust/issues/81395. There are no issues in our CI though, since it uses the nightly from 2021-01-07.

**Comment by jryans** [comment]
> I would hold merging this PR until the ICE is fixed.

The 2021-02-01 nightly includes a fix for this.

File: examples/dns/lib.rs [hunk 5]
```diff
@@ -59,7 +59,7 @@ mod dns {
     }
 
     /// Domain name service contract inspired by ChainX's [blog post]
-    /// (https://medium.com/@chainx_org/secure-and-decentralized-polkadot-domain-name-system-e06c35c2a48d).
+    /// (<https://medium.com/@chainx_org/secure-and-decentralized-polkadot-domain-name-system-e06c35c2a48d>).
```
Comment (Robbepop) [inline examples/dns/lib.rs:62]: What do these angle brackets change exactly?

File: crates/storage/src/collections/bitstash/mod.rs [hunk 6]
```diff
@@ -71,7 +71,8 @@ impl BitStash {
     /// in the `free` list.
     ///
     /// Returns the bit position of the first bit in the 256-bit chunk and not
-    /// the chunk position since that's what [`crate::collections::Bitvec::get_chunk`]
+    /// the chunk position since that's what
+    /// [`collections::Bitvec::get_chunk`][`crate::collections::Bitvec::get_chunk`]
```
Comment (Robbepop) [inline crates/storage/src/collections/bitstash/mod.rs:75]: I think in this case it is okay to also leave away `collections::` prefix and start with the type name.

Comment (Robbepop) [inline crates/storage/src/collections/bitstash/mod.rs:75]:
```suggestion
    /// the chunk position since that's what
    /// [`Bitvec::get_chunk`][`crate::collections::Bitvec::get_chunk`]
```

File: crates/storage/src/collections/smallvec/mod.rs [hunk 7]
```diff
@@ -89,7 +89,8 @@ where
     /// the contents of its associated storage region.
     ///
     /// This API is used for the `Drop` implementation of [`Vec`] as well as
-    /// for the [`crate::traits::SpreadLayout::clear_spread`] trait implementation.
+    /// for the [`traits::SpreadLayout::clear_spread`][`crate::traits::SpreadLayout::clear_spread`]
```
Comment (Robbepop) [inline crates/storage/src/collections/smallvec/mod.rs:92]:
```suggestion
    /// for the [`SpreadLayout::clear_spread`][`crate::traits::SpreadLayout::clear_spread`]
```

File: crates/storage/src/collections/stash/mod.rs [hunk 8]
```diff
@@ -266,7 +266,8 @@ where
     /// the contents of its associated storage region.
     ///
     /// This API is used for the `Drop` implementation of [`Vec`] as well as
-    /// for the [`crate::traits::SpreadLayout::clear_spread`] trait implementation.
+    /// for the [`traits::SpreadLayout::clear_spread`][`crate::traits::SpreadLayout::clear_spread`]
```
Comment (Robbepop) [inline crates/storage/src/collections/stash/mod.rs:269]:
```suggestion
    /// for the [`SpreadLayout::clear_spread`][`crate::traits::SpreadLayout::clear_spread`]
```

File: crates/storage/src/collections/vec/mod.rs [hunk 9]
```diff
@@ -109,7 +109,8 @@ where
     /// the contents of its associated storage region.
     ///
     /// This API is used for the `Drop` implementation of [`Vec`] as well as
-    /// for the [`crate::traits::SpreadLayout::clear_spread`] trait implementation.
+    /// for the [`traits::SpreadLayout::clear_spread`][`crate::traits::SpreadLayout::clear_spread`]
```
Comment (Robbepop) [inline crates/storage/src/collections/vec/mod.rs:112]:
```suggestion
    /// for the [`SpreadLayout::clear_spread`][`crate::traits::SpreadLayout::clear_spread`]
```

File: crates/storage/src/lazy/entry.rs [hunk 10]
```diff
@@ -182,7 +182,7 @@ where
     /// # Note
     ///
     /// Mainly used by lazy storage abstractions that only allow operating on
-    /// packed storage entities such as [`crate::lazy::LazyCell`].
+    /// packed storage entities such as [`lazy::LazyCell`][`crate::lazy::LazyCell`].
```
Comment (Robbepop) [inline crates/storage/src/lazy/entry.rs:185]:
```suggestion
    /// packed storage entities such as [`LazyCell`][`crate::lazy::LazyCell`].
```

File: crates/storage/src/lazy/entry.rs [hunk 11]
```diff
@@ -192,7 +192,7 @@ where
...
```
Comment (Robbepop) [inline crates/storage/src/lazy/entry.rs:195]:
```suggestion
    /// packed storage entities such as [`LazyCell`][`crate::lazy::LazyCell`].
```

File: crates/storage/src/lazy/entry.rs [hunk 12]
```diff
@@ -205,7 +205,7 @@ where
...
```
Comment (Robbepop) [inline crates/storage/src/lazy/entry.rs:208]:
```suggestion
    /// packed storage entities such as [`LazyCell`][`crate::lazy::LazyCell`].
```

File: crates/storage/src/lazy/entry.rs [hunk 13]
```diff
@@ -220,8 +220,8 @@ where
     /// # Note
     ///
     /// Mainly used by lazy storage abstractions that only allow operating on
-    /// packed storage entities such as [`crate::lazy::LazyIndexMap`] or
-    /// [`crate::lazy::LazyArray`].
+    /// packed storage entities such as [`lazy::LazyIndexMap`][`crate::lazy::LazyIndexMap`] or
+    /// [`lazy::LazyArray`][`crate::lazy::LazyArray`].
```
Comment (Robbepop) [inline crates/storage/src/lazy/entry.rs:224]:
```suggestion
    /// packed storage entities such as [`LazyIndexMap`][`crate::lazy::LazyIndexMap`] or
    /// [`LazyArray`][`crate::lazy::LazyArray`].
```

File: crates/storage/src/lazy/entry.rs [hunk 14]
```diff
@@ -231,8 +231,8 @@ where
     /// # Note
     ///
     /// Mainly used by lazy storage abstractions that only allow operating on
-    /// packed storage entities such as [`crate::lazy::LazyIndexMap`] or
-    /// [`crate::lazy::LazyArray`].
+    /// packed storage entities such as [`lazy::LazyIndexMap`][`crate::lazy::LazyIndexMap`]
+    /// or [`lazy::LazyArray`][`crate::lazy::LazyArray`].
```
Comment (Robbepop) [inline crates/storage/src/lazy/entry.rs:235]:
```suggestion
    /// packed storage entities such as [`LazyIndexMap`][`crate::lazy::LazyIndexMap`]
    /// or [`LazyArray`][`crate::lazy::LazyArray`].
```

File: crates/storage/src/lazy/entry.rs [hunk 15]
```diff
@@ -245,8 +245,8 @@ where
...
```
Comment (Robbepop) [inline crates/storage/src/lazy/entry.rs:249]:
```suggestion
    /// packed storage entities such as [`LazyIndexMap`][`crate::lazy::LazyIndexMap`]
    /// or [`LazyArray`][`crate::lazy::LazyArray`].
```

File: crates/storage/src/lazy/lazy_hmap.rs [hunk 16]
```diff
@@ -361,7 +361,8 @@ where
     ///
     /// This constructor is private and should never need to be called from
     /// outside this module. It is used to construct a lazy index map from a
-    /// key that is only useful upon a contract call. Use [`crate::lazy::LazyIndexMap::new`]
+    /// key that is only useful upon a contract call. Use
+    /// [`lazy::LazyIndexMap::new`][`crate::lazy::LazyIndexMap::new`]
```
Comment (Robbepop) [inline crates/storage/src/lazy/lazy_hmap.rs:365]:
```suggestion
    /// [`LazyIndexMap::new`][`crate::lazy::LazyIndexMap::new`]
```

**Review by Robbepop — COMMENTED** [review]
sorry for being so pedantic but simply accept my suggestions and we are good to merge

**Review by Robbepop — APPROVED** [review]
LGTM
